Extension/src/LanguageServer/configurations.ts

```diff
@@ -737,33 +738,57 @@ export class CppProperties {
         const result: string[] = [];
-        if (paths) {
-            paths = this.resolveDefaults(paths, defaultValue);
-            paths.forEach(entry => {
-                const entries: string[] = util.resolveVariables(entry, env).split(util.envDelimiter).map(e => this.resolvePath(e, false)).filter(e => e);
-                result.push(...entries);
-            });
-        }
-        if (!glob) {
+        if (paths === undefined) {
             return result;
         }
-
-        const globResult: string[] = [];
-        for (let res of result) {
-            // fastGlob will expand the ending double wildcard. temporary strip them before expanding
-            const recursive: boolean = res.endsWith('**');
-            if (recursive) {
-                res = res.slice(0, res.length - 2);
+        paths = this.resolveDefaults(paths, defaultValue);
+        paths.forEach(entry => {
+            const entries: string[] = util.resolveVariables(entry, env).split(util.envDelimiter).map(e => this.resolvePath(e, false)).filter(e => e);
+            if (!entry.includes('*')){
+                result.push(...entries);
+            }
+        });
+        if (glob) {
+            for (let res of paths) {
+                let counter: number = 0;
+                let slashFound: boolean = false;
+                const lastIndex: number = res.length - 1;
+                let isGlobPattern: boolean = false;
+                if (res.includes('*')){
+                    isGlobPattern = true;
+                }
+                // Detect and glob all wildcard variations by looking at last character in the path first.
+                for (let i: number = lastIndex; i >= 0; i--) {
+                    if (res[i] === '*') {
+                        counter++;
+                    } else if (res[i] === '/' || (res[i] === '\\' && isWindows)) {
+                        counter++;
+                        slashFound = true;
+                        break;
+                    } else {
+                        break;
+                    }
+                }
+                let suffix: string = '';
+                if (slashFound) {
+                    suffix = res.slice(res.length - counter);
+                    res = res.slice(0, res.length - counter);
+                }
+                let normalized: string = '';
+                let cwd: string = '';
+                if (isWindows) {
+                    normalized = res.replace(/\\/g, '/');
+                    cwd = this.rootUri?.fsPath?.replace(/\\/g, '/') || '';
+                }
+                // fastGlob silently strip non-found paths. limit that behavior to dynamic paths only
+                const matches: string[] = fastGlob.isDynamicPattern(normalized) ?
+                    fastGlob.sync(normalized, { onlyDirectories: true, cwd }) : [res];
+                if (isGlobPattern){
```

Comment:
Instead of `if (isGlobPattern)`, can it check `res.includes('*'))` directly instead of checking a var since `isGlobPattern` is only used here?

Reply:
Sorry, I didn't see `res` is getting modified after the check `res.includes('*')`.